Libraries/Lists/__tests__/VirtualizedList-test.js

```diff
@@ -506,161 +528,918 @@ describe('VirtualizedList', () => {
+  );
+
+  expect(component).toMatchSnapshot();
+});
+
+it('renders zero-height tail spacer on batch render if cells not yet measured and getItemLayout not defined', () => {
+  const items = generateItems(10);
+
+  let component;
+  ReactTestRenderer.act(() => {
+    component = ReactTestRenderer.create(
+      <VirtualizedList
+        initialNumToRender={3}
+        maxToRenderPerBatch={1}
+        windowSize={1}
+        {...baseItemProps(items)}
+      />,
+    );
+  });
+
+  ReactTestRenderer.act(() => {
+    simulateLayout(component, {
+      viewport: {width: 10, height: 50},
+      content: {width: 10, height: 200},
+    });
+    performNextBatch();
+  });
+
+  expect(component).toMatchSnapshot();
+});
+
+it('renders tail spacer up to last measured index if getItemLayout not defined', () => {
+  const items = generateItems(10);
+
+  let component;
+  ReactTestRenderer.act(() => {
+    component = ReactTestRenderer.create(
+      <VirtualizedList
+        initialNumToRender={3}
+        maxToRenderPerBatch={1}
+        windowSize={1}
+        {...baseItemProps(items)}
+      />,
+    );
+  });
+
+  ReactTestRenderer.act(() => {
+    const LAST_MEASURED_CELL = 6;
+    for (let i = 0; i <= LAST_MEASURED_CELL; ++i) {
+      simulateCellLayout(component, items, i, {
+        width: 10,
+        height: 10,
+        x: 0,
+        y: 10 * i,
+      });
+    }
+
+    simulateLayout(component, {
+      viewport: {width: 10, height: 50},
+      content: {width: 10, height: 30},
+    });
+    performNextBatch();
+  });
+
+  expect(component).toMatchSnapshot();
+});
+
+it('renders full tail spacer if all cells measured', () => {
+  const items = generateItems(10);
+
+  let component;
+  ReactTestRenderer.act(() => {
+    component = ReactTestRenderer.create(
+      <VirtualizedList
+        initialNumToRender={3}
+        maxToRenderPerBatch={1}
+        windowSize={1}
+        {...baseItemProps(items)}
+      />,
+    );
+  });
+
+  ReactTestRenderer.act(() => {
+    const LAST_MEASURED_CELL = 9;
+    for (let i = 0; i <= LAST_MEASURED_CELL; ++i) {
+      simulateCellLayout(component, items, i, {
+        width: 10,
+        height: 10,
+        x: 0,
+        y: 10 * i,
+      });
+    }
+
+    simulateLayout(component, {
+      viewport: {width: 10, height: 50},
+      content: {width: 10, height: 30},
+    });
+    performNextBatch();
+  });
+
+  expect(component).toMatchSnapshot();
+});
+
+it('renders tail spacer using frame average when getItemLayout undefined', () => {
+  const items = generateItems(10);
+
+  let component;
+  ReactTestRenderer.act(() => {
+    component = ReactTestRenderer.create(
+      <VirtualizedList
+        initialNumToRender={3}
+        maxToRenderPerBatch={1}
+        windowSize={1}
+        {...baseItemProps(items)}
+      />,
+    );
+  });
+
+  ReactTestRenderer.act(() => {
+    const LAST_MEASURED_CELL = 6;
+
+    let currentY = 0;
+    for (let i = 0; i <= LAST_MEASURED_CELL; ++i) {
```

Comment:
Need to rethink this test a bit. We're not using frame averages since we have cached cell layout for everything up to highest layout. Need to simulate cell unmounting.

Reply:
It looks like I was wrong about how this was supposed to work. We do keep all cached frames around, even after unmount. So we should I think always rely on existing measurements.